EX1/HW7.py

```diff
@@ -0,0 +1,373 @@
+        """String representation of the PostOffice.
+
+        :return: A string representation of the PostOffice.
+        :rtype: str
+        """
+        out = ''
+        for user, inbox in self.boxes.items():
+            out += f'{user}:\n'
+            if not inbox:
+                out += 'No messages.\n'
+            for msg in inbox:
+                out += f'Message {msg["id"]} from {msg["sender"]}: {msg["body"]}\n'
+        return out
+
+
+import random
+
+
+class Player2:
+    """
+    A class representing a player in a game.
+
+    Attributes:
+        name (str): The name of the player.
+        hp (int): The player's health points, initialized to 100.
+        exp (int): The player's experience points, initialized to 0.
+        level (int): The player's level, initialized to 1.
+        nemeses (list): A list of the player's enemies.
+
+    Methods:
+        attack(enemy_name=None):
+            Attacks the specified enemy or the last enemy in the list if no name is provided.
+            Returns True if the attack is successful, False otherwise.
+        revive():
+            Restores the player's health points to their initial value.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.hp = 100
+        self.exp = 0
+        self.level = 1
+        self.nemeses = []
+
+    def attack(self, enemy_name=None):
+        """
+        Attacks the specified enemy or the last enemy in the list if no name is provided.
+        Returns True if the attack is successful, False otherwise.
+        """
+        if not self.nemeses:
+            return False
+
+        if enemy_name:
+            for enemy in self.nemeses:
+                if enemy['name'] == enemy_name:
+                    damage = random.randint(5, 20) * self.level
+                    enemy['hp'] -= damage
+                    if enemy['hp'] <= 0:
+                        self.nemeses.remove(enemy)
+                    self.exp += damage
+                    return True
+        else:
+            enemy = self.nemeses[-1]
+            damage = random.randint(5, 20) * self.level
+            enemy['hp'] -= damage
+            if enemy['hp'] <= 0:
+                self.nemeses.remove(enemy)
+            self.exp += damage
+            return True
+
+        return False
+
+    def revive(self):
+        """
+        Restores the player's health points to their initial value.
+        """
+        self.hp = 100
+
+
+class Square:
+    def __init__(self, size):
+        self.size = size
+
+    def get_surface(self):
+        return self.size ** 2
+
+    def get_perimeter(self):
+        return 4 * self.size
+
+
+class Cube:
+    def __init__(self, size, color):
+        self.base = Square(size)
+        self.color = color
+        self.volume = self.base.get_surface() * size
+
+    def get_volume(self):
+        return self.volume
+
+
+class CubeTower:
+    def __init__(self):
+        self.cubes = []
+
+    def can_place_cube(self, cube):
```

Comment:
Use type hints and return types -
https://www.geeksforgeeks.org/duck-typing-in-python/
https://stackoverflow.com/questions/125367/dynamic-type-languages-versus-static-type-languages
https://docs.python.org/3/library/typing.html
https://realpython.com/lessons/pros-and-cons-type-hints/